packager/mux/mux.go

```diff
@@ -0,0 +1,93 @@
+package mux
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/ampproject/amppackager/packager/certcache"
+	"github.com/ampproject/amppackager/packager/signer"
+	muxp "github.com/ampproject/amppackager/packager/mux/params"
+	"github.com/ampproject/amppackager/packager/util"
+	"github.com/ampproject/amppackager/packager/validitymap"
+)
+
+type mux struct {
+	certCache *certcache.CertCache
+	signer *signer.Signer
+	validityMap *validitymap.ValidityMap
+}
+
+// Implements the HTTP routing strategy used by the amppkg server. The main
+// purpose for rolling our own is to be able to route the following types of
+// URLs:
+//   /priv/doc/https://example.com/esc%61ped%2Furl.html
+// and have the signer sign the URL exactly as it is encoded in the request, in
+// order to meet docs/cache_requirements.md.
+//
+// The default http mux had some problems which led to me investigating 3rd
+// party routers. (TODO(twifkak): Remember and document those problems.)
+//
+// I investigated two 3rd party routers capable of handling catch-all suffix
+// parameters:
+//   https://github.com/julienschmidt/httprouter
+//   https://github.com/dimfeld/httptreemux
+// Both libs unescape their catch-all parameters, making the above use-case
+// impossible. The latter does so despite documenting support for unmodified
+// URL escapings. This, plus a lack of feature needs, led me to believe that
+// writing our own mux was the best approach.
+func New(certCache *certcache.CertCache, signer *signer.Signer, validityMap *validitymap.ValidityMap) http.Handler {
+	return &mux{certCache, signer, validityMap}
+}
+
+func tryTrimPrefix(s, prefix string) (string, bool) {
+	sLen := len(s)
+	trimmed := strings.TrimPrefix(s, prefix)
+	return trimmed, len(trimmed) != sLen
+}
+
+var allowedMethods = map[string]bool{http.MethodGet: true, http.MethodHead: true}
+
+func (this *mux) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	if _, ok := allowedMethods[req.Method]; !ok {
+		http.Error(resp, "405 method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	params := map[string]string{}
+	req = muxp.WithParams(req, params)
+
+	// Use EscapedPath rather than RequestURI because the latter can take
+	// absolute-form, per https://tools.ietf.org/html/rfc7230#section-5.3.
+	//
+	// Use EscapedPath rather than RawPath to verify that the request URI
+	// is a valid encoding, per
+	// https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#application-signed-exchange
+	// item 3.
+	path := req.URL.EscapedPath()
+	if suffix, ok := tryTrimPrefix(path, "/priv/doc"); ok {
```

Comment:
Move "/priv/doc" into a global const var somewhere and dedupe it with main.go where it's also referenced. Or pass in "/priv/doc" as a parameter.

Reply:
It's not mentioned in main.go; this is the only production usage.

It's in signer_test.go, and I followed that same pattern with "/amppkg/cert" (certcache_test.go, signer_test.go) and "/amppkg/validity" (validitymap_test.go, signer_test.go). Vaguely thinking that it's better for tests to be explicit. I could change it if you think I should, but I'd do it for all, to be consistent.

Reply:
https://github.com/ampproject/amppackager/blob/releases/cmd/amppkg/main.go#L139 ?

Reply:
Removed in this PR.